wdl/downsampling_pipeline_part2.wdl

```diff
@@ -1,534 +1,638 @@
 
-        samtools index ~{downsample_file_sorted_name}
+    samtools index ~{downsample_file_sorted_name}
 
-        echo "File: indexed"        
+    echo "File: indexed"        
 
-    >>>
+  >>>
 
-    runtime {
-        docker: downsample_docker
-        memory: mem_size
-        cpu: "1"
-        disks: "local-disk " + disk_size + " HDD"
-        preemptible: 3
-        maxRetries: 1
-    }
+  runtime {
+    docker: downsample_docker
+    cpu: select_first([runtime_attr.cpu_cores, default_attr.cpu_cores])
+    memory: select_first([runtime_attr.mem_gb, default_attr.mem_gb]) + " GiB"
+    disks: "local-disk " + select_first([runtime_attr.disk_gb, default_attr.disk_gb]) + " HDD"
+    bootDiskSizeGb: select_first([runtime_attr.boot_disk_gb, default_attr.boot_disk_gb])
+    preemptible: select_first([runtime_attr.preemptible_tries, default_attr.preemptible_tries])
+    maxRetries: select_first([runtime_attr.max_retries, default_attr.max_retries])
+  }
 }
 
 task countCoverage {
     
-    input {
-        File downsample_sorted_cram
-        File reference_fasta
-        String downsample_docker
-        Int disk_size
-        String mem_size
-        File reference_dict
-        File ref_amb
-        File ref_ann
-        File ref_bwt
-        File ref_pac
-        File ref_sa
-        File ref_fai
-    }
-
-    String wgsCoverage_name = basename(downsample_sorted_cram, ".cram") + "_coverage.txt"
-    String ref_dictionary_name = "Homo_sapiens_assembly38.dict"
-
-    output {
-        File wgsCoverage = wgsCoverage_name
-    }
-
-    command {
-        java -Xmx32G -jar /opt/conda/share/picard-2.23.8-0/picard.jar CollectWgsMetrics \
-        I=~{downsample_sorted_cram} \
-        O=~{wgsCoverage_name} \
-        R=~{reference_fasta} \
-        COUNT_UNPAIRED=TRUE
-    }
-
-    runtime {
-        docker: downsample_docker
-        memory: mem_size
-        cpu: "1"
-        disks: "local-disk " + disk_size + " HDD"
-        preemptible: 3
-        maxRetries: 1
-    }
+  input {
+    File downsample_sorted_cram
+    File reference_fasta
+    String downsample_docker
+    File reference_dict
+    File ref_amb
+    File ref_ann
+    File ref_bwt
+    File ref_pac
+    File ref_sa
+    File ref_fai
+    RuntimeAttr? runtime_attr_override
+  }
+
+  Int num_cpu = 1
+  Int mem_size_gb = 4
+  Int vm_disk_size = 50
+
+  RuntimeAttr default_attr = object {
+    cpu_cores: num_cpu,
+    mem_gb: mem_size_gb, 
+    disk_gb: vm_disk_size,
+    boot_disk_gb: 10,
+    preemptible_tries: 3,
+    max_retries: 1
+  }
+
+  RuntimeAttr runtime_attr = select_first([runtime_attr_override, default_attr])
+
+  String wgs_coverage_name = basename(downsample_sorted_cram, ".cram") + "_coverage.txt"
+  String ref_dictionary_name = "Homo_sapiens_assembly38.dict"
+
+  output {
+    File wgs_coverage_file = wgs_coverage_name
+  }
+
+  command {
+    set -euo pipefail
+
+    java -Xmx32G -jar /opt/conda/share/picard-2.23.8-0/picard.jar CollectWgsMetrics \
```

Comment:
I notice that 32G is much larger than the default 4GB of memory given to this task. I know that 32G refers to the heap size which can include virtual memory, not just RAM, but wanted to check that this is correct?

Reply:
I believe you are correct. If I remember correctly, the VM kept failing because the heapsize ran out of space so I tried this and it worked. But I can cut down to see if it will still work and not fail